core/platform/bulk_email/mailchimp_bulk_email_services.py

```diff
@@ -0,0 +1,171 @@
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+import ast
+import hashlib
+
+import feconf
+import mailchimp3
+from mailchimp3 import mailchimpclient
+
+import python_utils
+
+
+def _get_subscriber_hash(email):
+    """Returns Mailchimp subscriber hash from email.
+
+    Args:
+        email: str. The email of the user.
+
+    Raises:
+        Exception. Invalid type for email, expected string.
+
+    Returns:
+        str. The subscriber hash corresponding to the input email.
+    """
+    if not isinstance(email, python_utils.BASESTRING):
+        raise Exception(
+            'Invalid type for email. Expected string, received %s' % email)
+    md5_hash = hashlib.md5()
+    md5_hash.update(email)
+    return md5_hash.hexdigest()
+
+
+def _get_mailchimp_class():
+    """Returns the mailchimp api class. This is separated into a separate
+    function to facilitate testing.
+
+    NOTE: No other functionalities should be added to this function.
+
+    Raises:
+        Exception. Mailchimp API key is not available.
+        Exception. Mailchimp username is not set.
+
+    Returns:
+        Mailchimp. A mailchimp class instance with the API key and username
+        initialized.
+    """
+    if not feconf.MAILCHIMP_API_KEY:
+        raise Exception('Mailchimp API key is not available.')
+
+    if not feconf.MAILCHIMP_USERNAME:
+        raise Exception('Mailchimp username is not set.')
+
+    return mailchimp3.MailChimp(
+        mc_api=feconf.MAILCHIMP_API_KEY, mc_user=feconf.MAILCHIMP_USERNAME)
+
+
+def permanently_delete_user_from_list(user_email):
+    """Permanently deletes the user with the given email from the Mailchimp
+    list.
+
+    NOTE: This should only be called from the wipeout service since once a user
+    is permanently deleted from mailchimp, they cannot be programmatically added
+    back via their API (the user would have to manually resubscribe back).
+
+    Args:
+        user_email: str. Email ID of the user. Email is used to uniquely
+            identify the user in the mailchimp DB.
+    """
+    client = _get_mailchimp_class()
+    subscriber_hash = _get_subscriber_hash(user_email)
+
+    try:
+        client.lists.members.get(
+            feconf.MAILCHIMP_AUDIENCE_ID, subscriber_hash)
+        client.lists.members.delete_permanent(
+            feconf.MAILCHIMP_AUDIENCE_ID, subscriber_hash)
+    except mailchimpclient.MailChimpError as error:
+        # This is has to be done since the message can only be accessed from
+        # MailChimpError by error.message in Python2, but this is deprecated in
+        # Python3.
+        # In Python3, the message can be accessed directly by KeyError
+        # (https://github.com/VingtCinq/python-mailchimp/pull/65), so as a
+        # workaround for Python2, the 'message' attribute is obtained by
+        # str() and then it is converted to dict.
```

Comment:
Explain (if true) that the current version works in both Py2 and Py3

Reply:
Done